### PR TITLE
Update to tf 2.8 and python 3.9 and drop s3fs dependency

### PR DESCRIPTION
**Pull request by ali-tny**
and generally relax some dependencies so we can actually use this downstream.

We remove the s3fs dependency in favour of using tentaclio, since the s3fs dependency was causing a lot of pain (strict requirements meant we couldn't update other packages)

### CODE REVIEW

File: dockerfiles/README.md [hunk 1]
```diff
@@ -1,7 +1,12 @@
-# Tensorflow 1.12 + Python 3.6 images
+# Tensorflow 1.12 + Python 3.9 images
```
Comment (ali-tny) [inline dockerfiles/README.md:1]: once merged I'll push the images to the ecr repo and update the description

File: timeserio/externals.py [hunk 2]
```diff
@@ -22,6 +22,13 @@ def __init__(self, name, *args, **kwargs):
         super().__init__(name, *args, **kwargs)
 
     def __getattr__(self, attr):
+        # Doctest attempts to unwrap everything to see if it can doctest it
+        # so we need to raise an AttributeError in that case rather than the
+        # specific ModuleNotFoundError
+        if attr == "__wrapped__":
```
Comment (ali-tny) [inline timeserio/externals.py:28]: not installing tensorpandas in dev deps lead to discovering this bug with NotFoundModules & doctest. we dont use tensorpandas anywhere else in tests anyway, and it restricts requirements a lot so it's good to get rid of it in the dev dependencies

File: tests/test_batches/test_chunked/test_parquet.py [hunk 3]
```diff
@@ -9,6 +13,27 @@
...
+        client = boto3.client("s3", "us-east-1")
+        client.create_bucket(Bucket=test_bucket_name)
+        url = f"s3://{test_bucket_name}"
+        yield url
+
+
+def _ddf_to_parquet(ddf, path):
+    for i, df in enumerate(ddf.partitions):
+        with tio.open(f"{path}/chunk_{i}.parquet", "wb") as fh:
+            df.compute().to_parquet(fh)
```
Comment (ali-tny) [inline tests/test_batches/test_chunked/test_parquet.py:34]: i noticed that timeserio was still depending on s3fs via the parquet batch generator that wasn't caught by tests, hence the additions here: adding to the tests to check writing to s3, and moving the mock data creation from s3fs to pure tentaclio

File: tests/test_version.py [hunk 4]
```diff
@@ -5,26 +5,40 @@
...
-                (?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)
-                (?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*
-            ))?
-            (?:\+(?P<build>
-                [0-9a-zA-Z-]+
-                (?:\.[0-9a-zA-Z-]+)*
-            ))?
-            $
-            """,
+        v?
```
Comment (capitancambio) [inline tests/test_version.py:13]: oh gosh you had to rewrite this regexp :scream: ?

File: timeserio/utils/pickle.py [hunk 5]
```diff
@@ -1,31 +1,18 @@
-import s3fs
+import tentaclio as tio
 import joblib.externals.cloudpickle as cp
 
 __all__ = ['dumpf', 'loadf']
 
 
-def open_url(filename, mode):
```
Comment (capitancambio) [inline timeserio/utils/pickle.py:7]: if I remember correctly, this exact function was the motivation to write tentaclio

Comment (capitancambio) [reply timeserio/utils/pickle.py:7]: the circle has been completed

**Review by capitancambio — APPROVED** [review]
this looks like a titanic effort, thanks. just out of curiosity, why changing from `relu` to `tanh` in the tests?

File: tests/test_version.py (hunk 4 above)
Comment (ali-tny) [reply tests/test_version.py:13]: noooo it's nicked from the python `packaging` package thank god

File: timeserio/utils/pickle.py (hunk 5 above)
Comment (ali-tny) [reply timeserio/utils/pickle.py:7]: ahaaaaa yep

**Comment by ali-tny** [comment]
@capitancambio changed to `tanh` since with the new random seed setting process it looked like things were getting randomly stuck with relu (ie in the test `test_fit_default` which checks that the metric after fitting 1 epoch of 13 random examples is lower than the starting values) - so moving it to tanh where it doesn't have the big zero section helped. nicked from igor's commit here: https://github.com/octoenergy/timeserio/pull/36/commits/bbc9ec739ff17af0e545e2a8866fb4e15f589db8

**Pull request merged**